### PR TITLE
[WIP] Support torch compile based pass manager framework

### PR DESCRIPTION
**Pull request by yuan-luo**
## Motivation

This PR is aimed to construct a torch compile based pass manager framework in SGLang. The idea and ground are from https://blog.vllm.ai/2025/08/20/torch-compile.html

In the end it will integrate the functionalities including:

**Piecewise CUDA Graphs**
Not all operations are compatible with CUDA Graphs; for example, [cascade attention is not](https://docs.vllm.ai/en/latest/design/v1/torch_compile.html#full-cudagraph-capture). SGLang works around this by breaking the captured graph into CUDA Graph -safe and -unsafe parts and executing them separately. This will give us the performance benefits of CUDA Graphs without losing correctness.

**Custom Compiler Passes in SGLang**
While torch.compile includes many built-in optimizations, SGLang adds custom compiler passes that apply additional optimizations to further improve performance. 

Fusion passes:
RMSNorm + Quant (FP8) fusion
SiLU-Mul + Quant (FP8) fusion
Attention + Quant (FP8) fusion (up to 7% improvement)
AllReduce + RMSNorm fusion (up to 15% improvement)
AllReduce + RMSNorm + Quant (FP8) fusion (up to 8% improvement)
AllReduce + RMSNorm + Quant (FP4) fusion (up to 10% improvement)
Sequence Parallelism & Async TP (up to 10% improvement)
Other passes:
No-op Elimination: eliminates or simplifies redundant reshape operations
Fix Functionalization: manually reinplaces auto_functionalized operations to avoid redundant copies and memory use

Most code pieces are adapted from vLLM torch compile framework with significant SGLang local customization since both frameworks are so different.

This is a WIP PR. Will refine and sharpen the code gradually. Will add vLLM credit in code references.

## Benchmarking and Profiling

Due to the PR is not ready, I benchmarked vLLM PIECEWISE CUDA GRAPH in prefill. 
DeepSeek-R1, TP8, 4k in 1 out. It seems PIECEWISE CUDA Graph feature does not perform very well in large model like DeepSeek-R1

```
vllm serve --model="deepseek-ai/DeepSeek-R1" --max-num-seqs 512 --trust-remote-code --tensor-parallel-size 8 --gpu-memory-utilization 0.9 --port 9256 --disable-log-requests --no-enable-prefix-caching -O '{"full_cuda_graph": true}' --cuda-graph-sizes 16 32 48 64 96 128 160 192 256 512 -O.cudagraph_mode=FULL_AND_PIECEWISE

vllm bench serve --backend openai \
    --base-url http://127.0.0.1:9256 \
    --dataset-name=random \
    --random-input-len 4000 --random-output-len 1 \
    --model deepseek-ai/DeepSeek-R1 \
    --seed 12345
```
With PIECEWISE CUDA GRAPH:
throughput 22151 tok/s
image

Close PIECEWISE CUDA GRAPH:
throughput 22113 tok/s
image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/compilation/collective_fusion.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+            "residual",
+            "norm_out",
+            "quant_out",
+            "scale_out",
+        ],
+        fake_impl=call_trtllm_fused_allreduce_norm_fake,
+    )
+    flashinfer_trtllm_fused_allreduce_norm = (
+        torch.ops.sglang.flashinfer_allreduce_residual_rmsnorm.default
+    )
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/compilation/collective_fusion.py]: Fixed.

File: python/sglang/srt/model_executor/model_runner.py [hunk 2]
```diff
@@ -1788,6 +1802,8 @@ def init_device_graphs(self):
             return
 
         if self.device != "cpu" and self.server_args.disable_cuda_graph:
+            if self.server_args.enable_torch_compile:
+                torch_compile(self.model, self.model_config)
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/model_executor/model_runner.py]: To do fix.

File: python/sglang/srt/model_executor/model_runner.py [hunk 3]
```diff
@@ -838,6 +844,14 @@ def load_model(self):
                 f"TP rank {self.tp_rank} could finish the model loading, but there are other ranks that didn't finish loading. It is likely due to unexpected failures (e.g., OOM) or a slow node."
             ) from None
 
+        if (
+            self.compilation_config.level == CompilationLevel.DYNAMO_AS_IS
+            and supports_dynamo()
+        ):
+            backend = self.compilation_config.init_backend(self.vllm_config)
+            self.model.compile(fullgraph=True, backend=backend)
+            return
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/model_executor/model_runner.py:868]: To do fix.

File: python/sglang/compilation/pass_manager.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                pass_(graph)
+                SglangInductorPass.dump_prefix += 1
+
+        # post-cleanup goes before fix_functionalization
+        # because it requires a functional graph
+        self.post_cleanup(graph)
+        SglangInductorPass.dump_prefix += 1
+
+        # always run fix_functionalization last
+        self.fix_functionalization(graph)
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/compilation/pass_manager.py]: to do fix.

File: python/sglang/compilation/collective_fusion.py [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+        global _FI_WORKSPACE_TENSOR
+        _FI_WORKSPACE_TENSOR = workspace_tensor
+        self.allreduce_params = FlashInferFusedAllReduceParams(
+            rank=rank,
+            world_size=self.tp_size,
+            use_fp32_lamport=use_fp32_lamport,
+            max_token_num=max_num_token,
+            # fuse rms norm static fp8 quant fused op
+            # in fallback path, when we don't use flashinfer
+            fuse_rms_quant=config.compilation_config.pass_config.enable_fusion,
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/compilation/collective_fusion.py:350]: to do fix.

[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/compilation/collective_fusion.py]: Fixed.

File: python/sglang/compilation/pass_manager.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/compilation/pass_manager.py]: Removed.

File: python/sglang/srt/configs/compilation_config.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    fi_allreduce_fusion_max_token_num: int = 16384
+    """Max number of tokens to used in flashinfer allreduce fusion."""
+
+    def uuid(self):
+        """
+        Produces a hash unique to the pass configuration.
+        Any new fields that affect compilation should be added to the hash.
+        Any future fields that don't affect compilation should be excluded.
+        """
+        return InductorPass.hash_dict(asdict(self))
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/configs/compilation_config.py:43]: Fixed.

File: python/sglang/srt/managers/tp_worker.py [hunk 7]
```diff
@@ -94,8 +96,17 @@ def __init__(
             tp_rank=tp_rank,
         )
 
+        # TODO(yuan-luo) hard-code compilation_config for POC,
+        # will refine in the former version.
+        self.compilation_config = CompilationConfig.from_server_args(
+            server_args,
+        )
+        self.device_config = DeviceConfig(self.model_runner.device, gpu_id)
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/managers/tp_worker.py]: Fixed.

File: python/sglang/srt/model_executor/model_runner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/model_executor/model_runner.py:868]: Fixed.

File: python/sglang/srt/model_executor/model_runner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/model_executor/model_runner.py]: Fixed.

File: python/sglang/srt/model_executor/cuda_graph_runner.py [hunk 8]
```diff
@@ -114,6 +114,23 @@ def _to_torch(model: torch.nn.Module, reverse: bool, num_tokens: int):
             _to_torch(sub, reverse, num_tokens)
 
 
+def _torch_compile_wrapper(forward):
+    return torch.compile(
+        torch.no_grad()(forward),
+        mode=os.environ.get("SGLANG_TORCH_COMPILE_MODE", "max-autotune-no-cudagraphs"),
+        dynamic=False,
+    )
```
Comment (eshoguli) [inline python/sglang/srt/model_executor/cuda_graph_runner.py:126]: You have to use `fullgraph=True`. It's merge stopper, isn't it?

Comment (yuan-luo) [reply python/sglang/srt/model_executor/cuda_graph_runner.py:126]: One of my PR's intentions is to resolve this issue. Will fix it.

**Comment by eshoguli** [comment]
Please, add `inductor` folder into `python/sglang/compilation` and move `collective_fusion.py`, `inductor_pass.py`, `pass_manager.py` and `sglang_inductor_pass.py` into `inductor` folder.

I'm going to support NPU https://github.com/sgl-project/sglang/pull/11104 with using `torch.fx.replace_pattern` but with the same configuration approach. I will add passes for `torch.fx.replace_pattern` into `python/sglang/compilation/fx` folder

**Comment by yuan-luo** [comment]
> Please, add `inductor` folder into `python/sglang/compilation` and move `collective_fusion.py`, `inductor_pass.py`, `pass_manager.py` and `sglang_inductor_pass.py` into `inductor` folder.
> 
> I'm going to support NPU #11104 with using `torch.fx.replace_pattern` but with the same configuration approach. I will add passes for `torch.fx.replace_pattern` into `python/sglang/compilation/fx` folder

@eshoguli Thanks for sharing this info. @DevashishLal-CB and I will proceed this task based on this #10987, we will work out a unified framework to achieve the pass manager including fusion, torch compile in compilation folder.

**Comment by eshoguli** [comment]
The same comment as for https://github.com/sgl-project/sglang/pull/10549:
@DevashishLal-CB and @yuan-luo , additional device specific change request: 
* devices: `cuda`, `npu` and `cpu` will add some transformations later,
* approaches which require different pass interface: `torch._inductor.pattern_matcher`, `torch.fx.replace_pattern` and manual `fx.Graph` traversal
* inference methods, possible they doesn't require separation but think about it: `CUDAGraph`, `NPUGraph`, `triton` for `CUDA` and `NPU`.

My suggestion is to make folder structure more general. Can, you, please move:
* general transformation options and general interfaces in `common` folder, 
* general passes in `common/passes`, cuda specific in `cuda/passes`,

What do you think about?

**Comment by yuan-luo** [comment]
Some duplication with https://github.com/sgl-project/sglang/pull/10062
will continue based on that PR.

**Comment by hnyls2002** [comment]
Thanks @yuan-luo! The torch-compile pass-manager work was superseded by the `compilation/` framework on
  main, so I'm closing this. Please reopen if something isn't covered there.
